src/clientHandler.ts

```diff
@@ -0,0 +1,252 @@
+import * as vscode from 'vscode';
+import ShortUniqueId from 'short-unique-id';
+import {
+	Executable,
+	LanguageClient,
+	LanguageClientOptions,
+	ServerOptions,
+	State,
+	DocumentSelector
+} from 'vscode-languageclient/node';
+import {
+	config,
+	getFolderName,
+	getWorkspaceFolder,
+	normalizeFolderName,
+	sortedWorkspaceFolders
+} from './vscodeUtils';
+import * as path from 'path';
+import TelemetryReporter from 'vscode-extension-telemetry';
+
+export interface TerraformLanguageClient {
+	commandPrefix: string,
+	client: LanguageClient
+}
+
+const clients: Map<string, TerraformLanguageClient> = new Map();
+
+/**
+ * ClientHandler maintains lifecycles of language clients
+ * based on the server's capabilities (whether multi-folder
+ * workspaces are supported).
+ */
+export class ClientHandler {
+	private shortUid: ShortUniqueId;
+	private pathToBinary: string;
+	private supportsMultiFolders = true;
+
+	constructor(private context: vscode.ExtensionContext, private reporter: TelemetryReporter ) {
+		this.shortUid = new ShortUniqueId();
+		this.pathToBinary = config('terraform').get('languageServer.pathToBinary');
+		if (this.pathToBinary) {
+			this.reporter.sendTelemetryEvent('usePathToBinary');
+		} else {
+			const installPath = path.join(context.extensionPath, 'lsp');
+			this.pathToBinary = path.join(installPath, 'terraform-ls');
+		}
+	}
+
+	public StartClients(folders?: string[]): vscode.Disposable[] {
+		const disposables: vscode.Disposable[] = [];
+
+		if (this.supportsMultiFolders) {
+			if (this.GetClient()?.client.needsStart()) {
+				console.log(`No need to start another client for ${folders}`)
+				return disposables;
+			}
+
+			console.log('Starting client');
+
+			const tfClient = this.newTerraformClient();
+			tfClient.client.onReady().then(async () => {
+				this.reporter.sendTelemetryEvent('startClient');
+				const multiFoldersSupported = tfClient.client.initializeResult.capabilities.workspace?.workspaceFolders?.supported;
+				console.log(`Multi-folder support: ${multiFoldersSupported}`);
+
+				if (!multiFoldersSupported) {
+					// restart is needed to launch folder-focused instances
```

Comment:
I'm probably lacking knowIedge here: Is this also happening if only a single workspace folder is used? (If yes: Is a restart required in that case?)

Reply:
No restart is needed because the client/server is already (optimistically) being launched with the right parameters for multi-folder. These parameters just aren't relevant for folder-specific instances. I tried to describe it more verbosely in the PR description above.